dlk/python/dlk/core/optimizer.py

```diff
@@ -14,632 +14,381 @@
-        self._check_and_transpose(node)
-
-
-class PreComputeRunner(GraphRunner):
-    """Optimization class that does precomputation and pruning on the graph.
-
-    Fron a constant node, this object precomputes as far as possible, and
-    replaces all precomputed nodes with a newly defined constant node.
-
-    Additionally, in the hard-quantized mode, this object replaces a
-    weight-quantizer node and succesive Conv node with a QConv node, and
-    packs the weight.
-    """
-
-    _quantized_bitwidth = 1
-    _wordsize = 32
-
-    def __init__(self, graph: Graph, hard_quantized: bool = False) -> None:
-        """Set up internal varibles."""
-        self._precomp_dic: Dict[str, bool] = {}
-        self._nodes_removed: Set[Operator] = set()
-        self._hard_quantized = hard_quantized
-        self._quantizers: Dict[str, Quantizer] = {}  # the operator name and its quantizer
-        self._connected_convs: Dict[Operator, List[Conv]] = {}  # node name and its connected convolver
-
-        super().__init__(graph)
-
-    def initialize(self, **kwargs: Any) -> None:
-        qconvs: List[Conv] = kwargs['qconv']
-        self._connected_convs = {q: [q] for q in qconvs}
+        to_be_removed.append(m.node)
 
-    def finalize(self, **kwargs: Any) -> None:
-        """Remove all unused nodes from the graph."""
-        for n in self._nodes_removed:
-            self._graph.remove_op(n)
+    for op in to_be_removed:
+        graph.remove_op(op)
 
-    # 1st phase: check which conv the node connects
 
-    def run_backward_by_default(self, node: Operator, **kwargs: Any) -> None:
-        outputs = node.output_op_list
+def pass_transpose(graph: Graph) -> None:
+    p = Pattern("*")
+    matches = find_pattern(graph, p)
 
-        convs: List[Conv] = sum([self._connected_convs[out] for out in outputs if self._connected_convs.get(out)], [])
-        self._connected_convs[node] = convs
+    for m in matches:
+        dim = m.node.dimension
+        shape = m.node.shape
+        if len(shape) != 4 or len(dim) != 4 or not set(dim).issubset({'N', 'H', 'W', 'C', 'I', 'O'}):
+            continue
 
-    def run_backward_conv(self, node: Conv, **kwargs: Any) -> None:
-        pass  # do nothing, as all (quantized) conv node is already registered to self._connected_convs
+        dim = dim.replace('I', 'C')
+        dim = dim.replace('O', 'N')
 
-    # 2nd phase: precompute and prune
+        permutation = list(map(lambda s: dim.index(s), 'NHWC'))
+        m.node.transpose(permutation)
 
-    def _has_precompute_value(self, op: Operator) -> bool:
-        """Return True if the operator has precompute value."""
-        return self._precomp_dic[op.name]
 
-    def _is_prunable(self, op: Operator) -> bool:
-        """Return True if op can be prunable."""
-        return self._has_precompute_value(op) and op.op_type != 'Constant'
+def pass_precompute(graph: Graph, processed_nodes) -> bool:
```

Comment:
Please write docstring all of pass

Reply:
Thank you, yes, documentation is needed.